# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

**Шаг 1.** Сформулируем гипотезу.

<u>**Гипотеза**</u>: возврат кредита в срок зависит от количества детей у клиента.

**Шаг 2.** Обратимся к датафрейму и определим, какие данные необходимо использовать для анализа.

In [22]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  
0               покупка жилья                     B  операции с недвижимостью  
1     приобретение автомобиля                     C    операции с автомобилем  
2               покупка жилья                     C  операции с недвижимостью  
3  дополнительное образование                     B     получение образования  
4             сыграть свадьбу                     C        проведение свадьбы

Для анализа подходят данные из 2 столбцов: `children` и `debt`. Из описания данных известно, что названия столбцов следующие:
* `children` — количество детей в семье;
* `debt` — имел ли задолженность по возврату кредитов

**Шаг 3.** Определим уникальные значения каждого из столбцов. Поскольку ранее была произведена предобработка данных всего датафрема, в дополнительных манипуляцих необходимости нет.

In [23]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

В столбце `children` представлены значения, которые легко интерпретируются, исходя из описания столбца в документации.

In [24]:
data['debt'].unique()

array([0, 1], dtype=int64)

В столбце `debt` представлены два значения: 0 и 1. В документации нет пояснения по поводу их интерпретации (необходимо уточнение), но, исходя из описания столбца, будем по умолчанию считать, что:
* 0 - клиент НЕ имел задолженность по возврату кредитов
* 1 - клиент имел задолженность по возврату кредитов

**Шаг 4.** Исследуем данные. Построим сводную таблицу с группировкой по столбцу `children`, где будет отображена информация об общем количестве клиентов по каждой группе и количестве клиентов, имевших задолженность по возврату кредитов:

In [25]:
pivot_children_debt = data.pivot_table(index='children', values='debt', aggfunc={'count','sum'})

# добавим дополнительный столбец 'share_debt' с расчётом процента клиентов-должников по каждой группе

pivot_children_debt['share_debtors'] = round(pivot_children_debt['sum'] / pivot_children_debt['count'] * 100, 2)
pivot_children_debt

count   sum  share_debtors
children                            
0         14091  1063           7.54
1          4808   444           9.23
2          2052   194           9.45
3           330    27           8.18
4            41     4           9.76
5             9     0           0.00

Из полученной выше таблицы мы видим, что доля клиентов с 3-мя и более детьми ничтожно мала по сравнению с другими группами, поэтому есть смысл категоризовать данные в следующие категории:

* **нет детей** - клиент не имеет детей;
* **1 ребёнок** - у клиента 1 ребёнок;
* **2 ребёнка** - у клиента 2 ребёнка;
* **3 и более детей** - у клиента 3 и более детей

Сделаем категоризацию и выведим обновлённую сводную таблицу, где доля должников будет отсортирована по возрастанию:

In [26]:
# напишем функцию children_category

def children_category(row):
    children = row['children']
    if children == 0:
        return 'нет детей'
    if children == 1:
        return '1 ребёнок'
    if children == 2:
        return '2 ребёнка'
    return '3 и более детей'

#напишем новый столбец 'children_category' в датафрейм

data['children_category'] = data.apply(children_category, axis=1)

In [27]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.

pivot_children_debt = data.pivot_table(index='children_category', values='debt', aggfunc={'count','sum'})
pivot_children_debt['share_debtors'] = round(pivot_children_debt['sum'] / pivot_children_debt['count'] * 100, 2)

# переименуем столбцы 'count' и 'sum'

pivot_children_debt = pivot_children_debt.rename(columns={'count': 'clients', 'sum': 'debtors'})

# отсортируем столбец 'share_debtors' по возрастанию

pivot_children_debt.sort_values(by='share_debtors', ascending=True)

clients  debtors  share_debtors
children_category                                 
нет детей            14091     1063           7.54
3 и более детей        380       31           8.16
1 ребёнок             4808      444           9.23
2 ребёнка             2052      194           9.45

**Вывод:** в целом можно подтвердить ранее сформулированную гипотезу: наблюдается незначительная корреляция между возвратом кредита в срок и количеством детей. Те клиенты, которые имеют хотя бы 1 ребёнка чаще нарушают срок возврата долга. Это можно объяснить тем, что наличие детей является причиной высокой финансовой нагрузки, из-за которой, в свою очередь, возникают просрочки. При этом стоит отметить, что, если рассматривать только клиентов, имеющих детей, то наименьший показатель доли приходится на клиентов с 3 и более детьми, а наибольший - на клиентов, у которых 2 ребёнка. Можно предположить, что семьи с 3 и более детьми более обеспечены в финансовом плане и уверены в своём будущем, если могут позволить себе такое многочисленное потомство и при этом вовремя платить по долгам. Но данная группа клиентов весьма малочисленна, в то время как клиенты, имеющих 2 ребёнка, - большая категория. Скорее всего, здесь представлены молодые люди, у которых присутствует желание быть счастливыми многодетными родителями, но пока их желание не до конца совпадает с их возможностями: текущая должность на работе не приносит необходимого дохода, а из-за чувства высокой ответственности перед своими детьми остальные заботы переходят на второй план, в т. ч. погашение кредитного долга.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

**Шаг 1.** Сформулируем гипотезу.

<u>**Гипотеза**</u>: возврат кредита в срок зависит от семейного положения клиента.

**Шаг 2.** Обратимся к датафрейму и определим, какие данные необходимо использовать для анализа **(см. п. 3.1)**

Для анализа подходят данные из 2 столбцов: `family_status` и `debt`. Из описания данных известно, что названия столбцов следующие:
* `family_status` — идентификатор семейного положения;
* `debt` — имел ли задолженность по возврату кредитов

**Шаг 3.** Определим уникальные значения каждого из столбцов. Поскольку ранее была произведена предобработка данных всего датафрема, в дополнительных манипуляцих необходимости нет.

In [28]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

В столбце `family_status` представлены значения, которые легко интерпретируются, исходя из описания столбца в документации. Но значение *'Не женат / не замужем'* начинается со строчной буквы. Переведём все значения в нижний регистр и снова выведем уникальные значения:

In [29]:
data['family_status'] = data['family_status'].str.lower()
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

Информацию по уникальным значениям столбца `debt` **см. п. 3.1.**

**Шаг 4.** Исследуем данные. Построим сводную таблицу с группировкой по столбцу `family_status`, где будет отображена информация об общем количестве клиентов по каждой группе и количестве клиентов, имевших задолженность по возврату кредитов:

In [30]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.

pivot_famstatus_debt = data.pivot_table(index='family_status', values='debt', aggfunc={'count','sum'})

# добавим дополнительный столбец 'share_debt' с расчётом процента клиентов-должников по каждой группе

pivot_famstatus_debt['share_debtors'] = round(pivot_famstatus_debt['sum'] / pivot_famstatus_debt['count'] * 100, 2)

# переименуем столбцы 'count' и 'sum'

pivot_famstatus_debt = pivot_famstatus_debt.rename(columns={'count': 'clients', 'sum': 'debtors'})

# отсортируем столбец 'share_debtors' по возрастанию

pivot_famstatus_debt.sort_values(by='share_debtors', ascending=True)

clients  debtors  share_debtors
family_status                                         
вдовец / вдова             951       63           6.62
в разводе                 1189       84           7.06
женат / замужем          12261      927           7.56
гражданский брак          4134      385           9.31
не женат / не замужем     2796      273           9.76

**Вывод:** гипотеза подтверждена. Действительно, между возвратом кредита в срок и семейным положением прослеживается корреляция. Самыми недобросовестными плательщиками являются клиенты из категорий *'не женат / не замужем'* и *'гражданский брак'.* Для выявления достоверных причин такого поведения необходимо дополнительное исследование, но можно предположить, что большую часть клиентов их этих категорий составляют молодые люди, у которых завышенные потребности, низкий доход и пониженное чувство ответственности, что в совокупности приводит к просрочкам. Средний показатель доли клиентов-должников наблюдается в категории *'женат / замужем',* на которую приходится больше всего выданных кредитов. Самый низкий показатель у клиентов категории *'вдовец / вдова',* где, скорее всего, представлены клиенты пожилого возраста с хорошим достатком, которые стремятся вовремя платить по долгам.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

**Шаг 1.** Сформулируем гипотезу.

<u>**Гипотеза**</u>: возврат кредита в срок зависит от уровня дохода клиента.

**Шаг 2.** Обратимся к датафрейму и определим, какие данные необходимо использовать для анализа **(см. п. 3.1)**

Для анализа подходят данные из 2 столбцов: `total_income_category` и `debt`. Из описания данных известно, что названия столбцов следующие:
* `total_income_category` — категории ежемесячного дохода (столбец сформирован на основании данных из столбца `total_income` — ежемесячный доход);
* `debt` — имел ли задолженность по возврату кредитов

**Шаг 3.** Определим уникальные значения каждого из столбцов. Поскольку ранее была произведена предобработка данных всего датафрема, в дополнительных манипуляцих необходимости нет.

In [31]:
data['total_income_category'].unique()

array(['B', 'C', 'D', 'E', 'A'], dtype=object)

В столбце `total_income_category` представлены значения, которые легко интерпретируются, исходя из описания столбца `total_income` в документации:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

Но при этом нам не известны денежные единицы данного дохода (требуется уточнение). По умолчанию будем считать, что ежемесячный доход выражен в <u>**рублях**<u/>.

Информацию по уникальным значениям столбца `debt` **см. п. 3.1.**

**Шаг 4.** Исследуем данные. Построим сводную таблицу с группировкой по столбцу `total_income_category`, где будет отображена информация об общем количестве клиентов по каждой группе и количестве клиентов, имевших задолженность по возврату кредитов:

In [32]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.

pivot_income_debt = data.pivot_table(index='total_income_category', values='debt', aggfunc={'count','sum'})

# добавим дополнительный столбец 'share_debt' с расчётом процента клиентов-должников по каждой группе

pivot_income_debt['share_debtors'] = round(pivot_income_debt['sum'] / pivot_income_debt['count'] * 100, 2)

# переименуем столбцы 'count' и 'sum'

pivot_income_debt = pivot_income_debt.rename(columns={'count': 'clients', 'sum': 'debtors'})

# отсортируем столбец 'share_debtors' по возрастанию

pivot_income_debt.sort_values(by='share_debtors', ascending=True)

clients  debtors  share_debtors
total_income_category                                 
D                          349       21           6.02
B                         5014      354           7.06
A                           25        2           8.00
C                        15921     1353           8.50
E                           22        2           9.09

**Вывод:** гипотеза подтверждена. Действительно, возврат кредита в срок зависит от ежемесячного дохода клиента, но при этом категории (доход соответсвующий категории) идут не по убыванию. Самыми добросовестными плательщиками оказались клиенты с доходом от 30 до 50 тыс. рублей. Это можно объяснить тем, что клиенты из данной категории стремятся вовремя платить по долгам, чтобы не обременять себя дополнительными расходами в виде пени, штрафов и иных санкций. Самая многочисленная категория клиентов с доходом от 50 до 200 тыс. рублей имеет один из самых высоких процентов просрочки. Здесь необходимо более детальное исследование для достоверного вывода, но, вероятно, в этой категории представлены клиенты, имеющие высокие затраты, с которыми не всегда удаётся справляться при среднем доходе. При этом вторая самая многочисленная категория клиентов с доходом от 200 тыс. до 1 млн. рублей имеет один самых низких рассматриваемых показателей. В данной группе, скорее всего, представлены клиенты, имеющие такой уровень заработка, который позволяет им не думать об экономии и вполне обеспеченно жить за счёт собственных средств. Но для крупных покупок они берут кредиты, и берут осознанно с полной уверенностью в том, что смогут погасить его в срок. Для этих клиентов кредит - это выгодный финансовый инструмент, позволяющый не только более рационально использовать денежные средства, но и получать материальные блага здесь и сейчас. Самую большую долю клиентов-должников составляют клиенты с самым низким доходом, что является логичным.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

**Шаг 1.** Сформулируем гипотезу.

<u>**Гипотеза**</u>: возврат кредита в срок зависит от цели его получения.

**Шаг 2.** Обратимся к датафрейму и определим, какие данные необходимо использовать для анализа **(см. п. 3.1)**

Для анализа подходят данные из 2 столбцов: `purpose_category` и `debt`. Из описания данных известно, что названия столбцов следующие:
* `purpose_category` — категории целей получения кредита (столбец сформирован на основании данных из столбца `purpose` — цель получения кредита);
* `debt` — имел ли задолженность по возврату кредитов

**Шаг 3.** Определим уникальные значения каждого из столбцов. Поскольку ранее была произведена предобработка данных всего датафрема, в дополнительных манипуляцих необходимости нет.

In [33]:
data['purpose_category'].unique()

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

В столбце `purpose_category` представлены значения, которые легко интерпретируются, исходя из описания столбца `purpose` в документации:

- операции с автомобилем,
- операции с недвижимостью,
- проведение свадьбы,
- получение образования.

Информацию по уникальным значениям столбца `debt` **см. п. 3.1.**

**Шаг 4.** Исследуем данные. Построим сводную таблицу с группировкой по столбцу `purpose_category`, где будет отображена информация об общем количестве клиентов по каждой группе и количестве клиентов, имевших задолженность по возврату кредитов:

In [34]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.

pivot_purpose_debt = data.pivot_table(index='purpose_category', values='debt', aggfunc={'count','sum'})

# добавим дополнительный столбец 'share_debt' с расчётом процента клиентов-должников по каждой группе

pivot_purpose_debt['share_debtors'] = round(pivot_purpose_debt['sum'] / pivot_purpose_debt['count'] * 100, 2)

# переименуем столбцы 'count' и 'sum'

pivot_purpose_debt = pivot_purpose_debt.rename(columns={'count': 'clients', 'sum': 'debtors'})

# отсортируем столбец 'share_debtors' по возрастанию

pivot_purpose_debt.sort_values(by='share_debtors', ascending=True)

clients  debtors  share_debtors
purpose_category                                         
операции с недвижимостью    10751      780           7.26
проведение свадьбы           2313      183           7.91
получение образования        3988      369           9.25
операции с автомобилем       4279      400           9.35

**Вывод:** гипотеза подтверждена. Действительно, возврат кредита в срок зависит от цели его получения. Клиенты, которые приобретают собственное жильё и стремятся к благополучию и созданию семьи, более ответственно относятся к уплате долгов, чтобы не возникало дополнительной финансовой нагрузки. У клиентов, выбирающих образование и автомобиль в качестве инвестиций, чаще возникают просрочки.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* в изначальном датафрейме пропуски возникли в 2 столбцах: `days_employed` и `total_income`. Пропусков было одинаковое количество - 2174 ед., из чего можно сделать вывод, что заполнение одного столбца влияло на заполнение другого или наоборот. Причины пропусков: 
* необязательные данные для заполнения (т.е. клиент пожелал оставить эти поля пустыми;
* технический сбой (маловероятно)

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Среднее арифметическое и медиана оценивают значения в центре выборки. Когда есть оторванные от основной массы "показатели-лидеры", их результаты сильно смещают значение среднего арифметического вверх. Поэтому оценивать показатели лучше медианой.

### Шаг 4: общий вывод.

Для кредитного отдела банка (делее заказчик) было проведено исследование надёжности заёмщиков, цель которого состояла в выявлении влияния определённых показателей на факт погашения кредита в срок. 

**Для анализа заказчик предоставил файл в CSV-формате со статистикой о платёжеспособности клиентов. В файле были представлены следующие данные**:

* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id`— идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

На основании полученной информации необходимо было разобраться, существует ли зависимость между показателем `debt` (имел ли задолженность по возврату кредитов) и показателями:

* `children` — количество детей в семье
* `family_status` — семейное положение
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

**Перед анализом была проведена предобработка дынных. На этой стадии выполнены следующие действия**:

1. Удалены пропуски в столбцах `days_employed` и `total_income`: пропуски заполнены медианным значением;
2. Произведена обработка аномальных значений (артефактов):

   * отрицательные значения заменены положительными в столбце `days_employed`;
   * удалены артефакты в столбце `children`;
   
   
3. Заменён тип данных в столбце `total_income` с вещественного на целочисленный;
4. Обработаны неявные дубликаты;
5. Выполнена категоризация данных столбцов `total_income`, `purpose` и `children`. В датафрейм добавлены новые столбцы: `total_income_category`, `purpose_category` и `children_category`.

**Далее был проведён анализ: сформулированы и проверены гипотезы, которые позволили сделать следующий вывод**:
    
Прослеживается невысокая корреляция между рассматриваемыми показателями, но этого достаточно, чтобы обрисовать предположительные портреты надежного и рискового заемщиков. 

С наибольшей вероятностью просрочку допустят те клиенты, которые не состоят в браке (или состоят в гражданском браке), имеют доход менее 30 тыс. рублей, но при этом являются родителем 1-2 детей. Как правило, основная цель получения кредита у данной группы клиентов - это операции с автомобилем или получение образования.

Надёжными заёмщиками можно считать клиентов, не имеющих детей и пребывающие в плане семейного положения в статусе вдовца или вдовы, либо в статусе разведённого мужчины или разведённой женщины, иными словами являющимися одиночками. В основном они берут кредиты для операций с недвижимостью и проведения свадьбы, так как у них нет возможности совершать крупные приобретения товаров/услуг за счёт собственных средств по причине их низкого дохода (от 30 до 50 тыс. рублей). Но при этом эти люди ответственно подходят к уплате долгов.

Результаты проведённого исследования можно использовать для построения модели кредитного скоринга, но стоит учитывать следующее: существует вероятность того, что не все вышеизложенные факторы могут совпасть, а лишь только их часть. Одной из причин этому могут послужить пропущенные значения в предоставленной таблице с данными (в т.ч. в ключевом показателе для анализа - ежемесячном доходе). В любом случае комбинация нескольких факторов может указывать на возможные кредитные просрочки в будущем.